Fix: object-shorthand providing invalid fixes for typescript

bradzacher opened the pull request:
<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

**What is the purpose of this pull request? (put an "X" next to item)**

[x] Bug fix 

Related: https://github.com/typescript-eslint/typescript-eslint/issues/124

The bug is that currently the fixer adds extra parentheses when fixing a TS arrow function with a return type:

```ts
/* eslint object-shorthand: [1, "always", { avoidExplicitReturnArrows: true }] */

const test = {
  key: (arg: any): any => {}
}

///// fixes to

const test = {
  key((arg: any): any) {}
}
```

This happens because the rule currently searches backwards from the `=>` to find a `)`, to determine if it should wrap the arguments in parentheses. It obviously instead finds a return type annotation, hence the breakage.

**What changes did you make? (Give an overview)**

Adjusted the fix algorithm for arrow functions so that it searches from the start of the arrow function expression instead.

I added a TS-specific test for this - happy to remove it if you don't want the burden.

platinumazure commented on lib/rules/object-shorthand.js:
```diff
@@ -264,10 +264,13 @@ module.exports = {
-            const hasParensAroundParameters = tokenBeforeArrow.type === "Punctuator" && tokenBeforeArrow.value === ")";
-            const oldParamText = sourceCode.text.slice(sourceCode.getFirstToken(node.value, node.value.async ? 1 : 0).range[0], tokenBeforeArrow.range[1]);
+
+            const firstToken = sourceCode.getFirstToken(node.value, node.value.async ? 1 : 0);
```
I'm a little concerned that this could be brittle if more arrow function modifiers are added in future language versions.

I'm wondering if it would be possible to search for tokens between the first token of `node.value` (inclusive), and the first token of `node.value.params[0]` (exclusive)? `sourceCode.getTokensBetween()` does an exclusive search, so maybe it could be something like `sourceCode.getTokensBetween(node.key, node.value.params[0], { filter: astUtils.isOpenParenToken })`?

If it's too much of a pain, I'm okay with what you have now, but I wanted to raise the question.

bradzacher commented:
I've refactored the code so it doesn't rely on a check for `async` to get the first token.

I've also added a test with typescript code to help prevent a regression.
